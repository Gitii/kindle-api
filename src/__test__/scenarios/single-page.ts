import { Scenario } from "../scenario";
import { sample1 } from "./fixtures/books";
import { defaultRequestBody } from "./fixtures/default-request-body";
import { sessionToken } from "./fixtures/tokens";

export const singlePages = {
  firstPage: {
    request: {
      method: "post",
      url: "http://localhost:8080/api/forward",
      body: {
        ...defaultRequestBody,
        requestUrl:
          "https://read.amazon.com/kindle-library/search?query=&libraryType=BOOKS&sortType=acquisition_desc&querySize=50",
        headers: {
          ...defaultRequestBody.headers,
          "x-amzn-sessionid": "2",
          "x-adp-session-token": sessionToken,
        },
      },
    },
    response: {
      headers: {},
      status: 200,
      body: {
        cookies: {
          "session-id": "2",
        },
        body: JSON.stringify({
          itemsList: [sample1],
        }),
      },
      meta: {
        books: [sample1],
      },
    },
  },
} satisfies Scenario;
