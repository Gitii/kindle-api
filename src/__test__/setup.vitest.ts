import { beforeAll, afterEach, afterAll } from "vitest";
import { server } from "./setup.msw";

beforeAll(() =>
  server.listen({
    onUnhandledRequest: "error",
  })
);
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
