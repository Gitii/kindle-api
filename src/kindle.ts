import { KindleBook, KindleBookData } from "./book.js";
import { KindleRequiredCookies, HttpClient, TLSClient } from "./http-client.js";

export type { KindleBook, KindleBookDetails } from "./book.js";
export type { KindleOwnedBookMetadataResponse } from "./book-metadata";

export type KindleConfiguration = {
	/**
	 * Cookie string copied from your browser or exact
	 * requried cookies
	 */
	cookies: KindleRequiredCookies | string;
	deviceToken: string;
	// Optional
	clientVersion?: string;
};

export type KindleOptions = {
	config: KindleConfiguration;
	sessionId: string;
};

export type KindleFromCookieOptions = {
	cookieString: string;
	deviceToken: string;
};

export class Kindle {
	public static DEVICE_TOKEN_URL =
		"https://read.amazon.com/service/web/register/getDeviceToken";
	public static readonly BOOKS_URL =
		"https://read.amazon.com/kindle-library/search?query=&libraryType=BOOKS&sortType=recency&querySize=50";

	/**
	 * The default list of books fetched when setting up {@link Kindle}
	 *
	 * We need to hit up the books endpoint to get necessary cookies
	 * so we save the initial response here just to make sure the
	 * user doesn't have to run the same request twice for no reason
	 */
	readonly defaultBooks: KindleBook[];
	readonly #client: HttpClient;

	constructor(
		private options: KindleOptions,
		client: HttpClient,
		// not necessary for initialization (if called from the outside)
		// so we're leaving this nullable
		prePopulatedBooks?: KindleBook[],
	) {
		this.defaultBooks = prePopulatedBooks ?? [];
		this.#client = client;
	}

	/**
	 * Destroys the underlying go server.
	 */
	async destroy() {
		await TLSClient.destroy();
	}

	static async fromConfig(config: KindleConfiguration): Promise<Kindle> {
		const cookies =
			typeof config.cookies === "string"
				? Kindle.deserializeCookies(config.cookies)
				: config.cookies;
		const client = new HttpClient(cookies);

		const { sessionId, books } = await Kindle.baseRequest(client);
		client.updateSession(sessionId);
		const deviceInfo = await Kindle.deviceToken(client, config.deviceToken);
		client.updateAdpSession(deviceInfo.deviceSessionToken);

		return new this(
			{
				config: { cookies, deviceToken: config.deviceToken },
				sessionId,
			},
			client,
			books,
		);
	}

	static async deviceToken(
		client: HttpClient,
		token: string,
	): Promise<KindleDeviceInfo> {
		const params = new URLSearchParams({
			serialNumber: token,
			deviceType: token,
		});
		const url = `${Kindle.DEVICE_TOKEN_URL}?${params.toString()}`;
		const response = await client.request(url);
		return response.body as KindleDeviceInfo;
	}

	static async baseRequest(
		client: HttpClient,
		version?: string,
	): Promise<{
		books: KindleBook[];
		sessionId: string;
	}> {
		type Response = {
			itemsList: KindleBookData[];
		};

		const response = await client.request(Kindle.BOOKS_URL);
		const newCookies = client.extractSetCookies(response);
		const sessionId = newCookies["session-id"];

		return {
			books: (response.body as Response).itemsList.map(
				(book) => new KindleBook(book, client, version),
			),
			sessionId,
		};
	}

	async books(): Promise<KindleBook[]> {
		const result = await Kindle.baseRequest(this.#client);
		// refreshing the internal session every time books is called.
		// This doesn't prevent us from calling the books endpoint but
		// it does prevent requesting the metadata of individual books
		this.options.sessionId = result.sessionId;
		return result.books;
	}

	static deserializeCookies(cookies: string): KindleRequiredCookies {
		const values = cookies
			.split(";")
			.map((v) => v.split("="))
			.reduce((acc, [key, value]) => {
				acc[decodeURIComponent(key.trim())] = decodeURIComponent(value.trim());
				return acc;
			}, {} as Record<string, string>);

		return {
			atMain: values["at-main"],
			sessionId: values["session-id"],
			ubidMain: values["ubid-main"],
		};
	}
}

export interface KindleDeviceInfo {
	clientHashId: string;
	deviceName: string;
	deviceSessionToken: string;
	eid: string;
}
