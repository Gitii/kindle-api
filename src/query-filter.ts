export interface Query {
  /**
   * sortType
   */
  sortType:
    | "recency"
    | "title"
    | "author"
    | "acquisition_desc"
    | "acquisition_asc";
}

export interface Filter {
  /**
   * Filter by abonnement.
   */
  originType?: "KINDLE_UNLIMITED" | "PRIME" | "COMICS_UNLIMITED";
  /**
   * Request the next page of results.
   */
  paginationToken?: string;
  /**
   * Set the number of results to return.
   * Default is 50.
   */
  querySize?: number;
  /**
   * When false, pagination is enabled and all results are returned.
   * The default is false and only the first page of results is returned.
   */
  firstPageOnly?: boolean;
}
