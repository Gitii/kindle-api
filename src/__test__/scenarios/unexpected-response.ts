import { TLSClientResponseData } from "../../kindle";
import { Scenario } from "../scenario";
import { defaultRequestBody } from "./fixtures/default-request-body";

export function unexpectedResponse({
  response,
}: {
  response: TLSClientResponseData;
}) {
  return {
    startSession: {
      request: {
        method: "post",
        url: "http://localhost:8080/api/forward",
        body: {
          ...defaultRequestBody,
          requestUrl:
            "https://read.amazon.com/kindle-library/search?query=&libraryType=BOOKS&sortType=acquisition_desc&querySize=50",
        },
      },
      response: {
        headers: {},
        status: 200,
        body: response,
        meta: {
          response,
        },
      },
    },
  } satisfies Scenario;
}
