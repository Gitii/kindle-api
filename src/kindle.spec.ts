import "dotenv/config";
import { test, expect, describe } from "vitest";
import { Kindle, KindleConfiguration, TLSClientResponseData } from "./kindle";
import { useScenario } from "./__test__/scenario";
import { singleBook } from "./__test__/scenarios/single-book";
import { multiplePages } from "./__test__/scenarios/multiple-pages";
import { startSession } from "./__test__/scenarios/start-session";
import { signinRedirect } from "./__test__/scenarios/signin-redirect";
import { getError } from "./__test__/get-error";
import { Filter } from "./query-filter";
import { AuthSessionError } from "./errors/auth-session-error";
import { unexpectedResponse } from "./__test__/scenarios/unexpected-response";
import { UnexpectedResponseError } from "./errors/unexpected-response-error";

const cookies = process.env.COOKIES;

function config(): KindleConfiguration {
  const deviceToken = process.env.DEVICE_TOKEN;
  if (!(deviceToken && cookies)) {
    throw Error("Invalid configuration");
  }
  return {
    deviceToken,
    cookies,
    tlsServer: {
      // rome-ignore lint/style/noNonNullAssertion: <explanation>
      apiKey: process.env.TLS_SERVER_API_KEY!,
      // rome-ignore lint/style/noNonNullAssertion: <explanation>
      url: process.env.TLS_SERVER_URL!,
    },
  };
}

test("should fetch first page when created", async () => {
  // given
  const { getBookDetails } = useScenario(singleBook);
  useScenario(startSession({ books: getBookDetails.meta.books }), {
    append: true,
  });

  // when
  const kindle = await Kindle.fromConfig(config());

  // then
  expect(kindle.defaultBooks).toMatchSnapshot();
  expect(await kindle.defaultBooks[0].fullDetails()).toMatchSnapshot();
});

describe("pagination", () => {
  test.each([
    {
      fetchAllPages: undefined,
    },
    {
      fetchAllPages: false,
    },
  ] satisfies Filter[])(
    "should only get the first page of book results when %s",
    async (filter) => {
      // given
      useScenario(startSession()); // used for initial setup
      const kindle = await Kindle.fromConfig(config());
      expect(kindle.defaultBooks.length).toBe(0);

      useScenario(multiplePages); // used for this test

      // when
      const books = await kindle.books({ filter });

      // then
      expect(books).toMatchSnapshot();
      expect(books.length).toBe(1);
    }
  );

  test("should fetch all books when fetchAllPages=true", async () => {
    // given
    useScenario(startSession()); // used for initial setup
    const kindle = await Kindle.fromConfig(config());
    expect(kindle.defaultBooks.length).toBe(0);

    useScenario(multiplePages); // used for this test

    // when
    const books = await kindle.books({ filter: { fetchAllPages: true } });

    // then
    expect(books).toMatchSnapshot();
    expect(books.length).toBe(2);
  });
});

describe("auth errors", () => {
  test("should throw when session expired", async () => {
    // given
    const {
      startSession: {
        meta: { response },
      },
    } = useScenario(signinRedirect());

    // when
    const error = await getError(
      async (): Promise<unknown> => await Kindle.fromConfig(config())
    );

    // then
    expect(error).toBeInstanceOf(AuthSessionError);
    expect(error).toEqual(
      expect.objectContaining({
        message: "Session expired",
        response,
      })
    );
  });
});

describe("unexpected response errors", () => {
  test.each([400])(
    "should throw when response status is unexpected %s",
    async (status) => {
      // given
      const response = {
        headers: {},
        status,
        body: "{}",
        cookies: {},
        target:
          "https://read.amazon.com/kindle-library/search?query=&libraryType=BOOKS&sortType=acquisition_desc&querySize=50",
      } satisfies TLSClientResponseData;
      useScenario(unexpectedResponse({ response }));

      // when
      const error = await getError(
        async (): Promise<unknown> => await Kindle.fromConfig(config())
      );

      // then
      expect(error).toBeInstanceOf(UnexpectedResponseError);
      expect(error).toEqual(
        expect.objectContaining({
          message: `Unexpected status code: ${status}`,
          response,
        })
      );
    }
  );
});
