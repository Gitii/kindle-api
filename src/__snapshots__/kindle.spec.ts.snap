// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pagination > should fetch all books when fetchAllPages=true 1`] = `
[
  KindleBook {
    "asin": "B000FBJG4U",
    "authors": [
      {
        "firstName": "Jules Verne",
        "lastName": "",
      },
    ],
    "imageUrl": "https://www.amazon.com/dp/B000FBJG4U",
    "mangaOrComicAsin": false,
    "originType": "purchased",
    "productUrl": "https://www.amazon.com/dp/B000FBJG4U",
    "resourceType": "EBOOK",
    "title": "Journey to the Center of the Earth",
    "webReaderUrl": "https://read.amazon.com/kp/kshare?asin=B000FBJG4U&id=wRj3DwAAQBAJ",
  },
  KindleBook {
    "asin": "B08R3YVH53",
    "authors": [
      {
        "firstName": "F. Scott Fitzgerald",
        "lastName": "",
      },
    ],
    "imageUrl": "https://www.amazon.com/dp/B08R3YVH53",
    "mangaOrComicAsin": false,
    "originType": "sampled",
    "productUrl": "https://www.amazon.com/dp/B08R3YVH53",
    "resourceType": "EBOOK_SAMPLE",
    "title": "The Great Gatsby",
    "webReaderUrl": "https://read.amazon.com/kp/kshare?asin=B08R3YVH53&id=wRj3DwAAQBAJ",
  },
]
`;

exports[`pagination > should only get the first page of book results when { fetchAllPages: false } 1`] = `
[
  KindleBook {
    "asin": "B000FBJG4U",
    "authors": [
      {
        "firstName": "Jules Verne",
        "lastName": "",
      },
    ],
    "imageUrl": "https://www.amazon.com/dp/B000FBJG4U",
    "mangaOrComicAsin": false,
    "originType": "purchased",
    "productUrl": "https://www.amazon.com/dp/B000FBJG4U",
    "resourceType": "EBOOK",
    "title": "Journey to the Center of the Earth",
    "webReaderUrl": "https://read.amazon.com/kp/kshare?asin=B000FBJG4U&id=wRj3DwAAQBAJ",
  },
]
`;

exports[`pagination > should only get the first page of book results when { fetchAllPages: undefined } 1`] = `
[
  KindleBook {
    "asin": "B000FBJG4U",
    "authors": [
      {
        "firstName": "Jules Verne",
        "lastName": "",
      },
    ],
    "imageUrl": "https://www.amazon.com/dp/B000FBJG4U",
    "mangaOrComicAsin": false,
    "originType": "purchased",
    "productUrl": "https://www.amazon.com/dp/B000FBJG4U",
    "resourceType": "EBOOK",
    "title": "Journey to the Center of the Earth",
    "webReaderUrl": "https://read.amazon.com/kp/kshare?asin=B000FBJG4U&id=wRj3DwAAQBAJ",
  },
]
`;

exports[`should fetch first page when created 1`] = `
[
  KindleBook {
    "asin": "B000FBJG4U",
    "authors": [
      {
        "firstName": "Jules Verne",
        "lastName": "",
      },
    ],
    "imageUrl": "https://www.amazon.com/dp/B000FBJG4U",
    "mangaOrComicAsin": false,
    "originType": "purchased",
    "productUrl": "https://www.amazon.com/dp/B000FBJG4U",
    "resourceType": "EBOOK",
    "title": "Journey to the Center of the Earth",
    "webReaderUrl": "https://read.amazon.com/kp/kshare?asin=B000FBJG4U&id=wRj3DwAAQBAJ",
  },
]
`;

exports[`should fetch first page when created 2`] = `
{
  "asin": "B000FBJG4U",
  "authors": [
    {
      "firstName": "Jules Verne",
      "lastName": "",
    },
  ],
  "bookType": "owned",
  "coverUrl": "https://www.amazon.com/dp/B000FBJG4U",
  "endPosition": 100,
  "formatVersion": "1.0",
  "largeCoverUrl": "https://www.amazon.com/dp/B000FBJG4U",
  "mangaOrComicAsin": false,
  "metadataUrl": "https://read.amazon.com/service/metadata/lookup?asin=B000FBJG4U&metadataVersion=1",
  "originType": "purchased",
  "percentageRead": 1,
  "productUrl": "https://www.amazon.com/dp/B000FBJG4U",
  "progress": {
    "position": 0,
    "reportedOnDevice": "Kindle Reader",
    "syncDate": 1970-01-01T00:00:00.000Z,
  },
  "publisher": "Public Domain Books",
  "releaseDate": "2009-04-01T00:00:00.000Z",
  "srl": 0,
  "startPosition": 1,
  "title": "Journey to the Center of the Earth",
  "webReaderUrl": "https://read.amazon.com/kp/kshare?asin=B000FBJG4U&id=wRj3DwAAQBAJ",
}
`;
